### PR TITLE
Add basic fake NTLM server to test NTAuthentication round-trip scenarios

### PR DESCRIPTION
**Pull request by filipnavara**
Add basic low-level NTLM exchange test. This is prerequisite for #29270, #62264 and test for the fixed scenario in #54101. It can also serve as basis for future tests for specific issues.

### CODE REVIEW

**Comment by filipnavara** [comment]
Well, that is interesting, looks like NTLMSSP on Linux doesn't enable message integrity unless it's enforced through the SPNEGO protocol.

**Comment by filipnavara** [comment]
Putting back to draft. The test data in the specification use incorrect flags in the example which results in failing exchange.

**Comment by filipnavara** [comment]
Getting one more failure locally, on Linux the SPN is not propagated.

File: src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                ContextFlagsPal.Connection | ContextFlagsPal.InitIntegrity, null);
+
+            DoNtlmExchange(fakeNtlmServer, ntAuth);
+
+            Assert.True(fakeNtlmServer.IsAuthenticated);
+            // NTLMSSP on Linux doesn't send the MIC and sends incorrect SPN (drops the service prefix)
+            if (!OperatingSystem.IsLinux())
+            {
+                Assert.True(fakeNtlmServer.IsMICPresent);
+                Assert.Equal("HTTP/foo", fakeNtlmServer.ClientSpecifiedSpn);
```
Comment (filipnavara) [inline src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs:90]: [gss-ntlmssp seems to throw away the service name](https://github.com/gssapi/gss-ntlmssp/blob/9b6493b1419d615e347d67c636fd238aa4d0f780/src/gss_names.c#L297-L319). I think that's a bug in their implementation. Neither Windows nor macOS seems to do that.

File: src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                isServer: false, "NTLM", s_testCredentialRight, "HTTP/foo",
+                ContextFlagsPal.Connection | ContextFlagsPal.InitIntegrity, null);
+
+            DoNtlmExchange(fakeNtlmServer, ntAuth);
+
+            Assert.True(fakeNtlmServer.IsAuthenticated);
+            // NTLMSSP on Linux doesn't send the MIC and sends incorrect SPN (drops the service prefix)
+            if (!OperatingSystem.IsLinux())
+            {
+                Assert.True(fakeNtlmServer.IsMICPresent);
```
Comment (filipnavara) [inline src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs:89]: The MIC is currently not sent due to [this code branch](https://github.com/gssapi/gss-ntlmssp/blob/9b6493b1419d615e347d67c636fd238aa4d0f780/src/gss_auth.c#L77-L83).

It seems like there may be a way to enforce it by calling `gss_inquire_sec_context_by_oid` with the [`1.3.6.1.4.1.7165.655.1.2` OID](https://github.com/gssapi/gss-ntlmssp/blob/c0a02166f20dac6a1712aaf2fd15757a02f8586d/src/gssapi_ntlmssp.h#L44) and one-byte data buffer. The provider implements it [here](https://github.com/gssapi/gss-ntlmssp/blob/1cbc124fc15d1fad1e34e885f98e449da2dbcf21/src/gss_sec_ctx.c#L1218-L1251).

**Comment by filipnavara** [comment]
I have tried to make a test for `NTAuthentication.MakeSignature` only to find out that it produces different output on macOS and Linux. On macOS the result was (content, signature) and on Linux it was (signature, content) in the output buffer. I am not sure yet which one is wrong but there is a LOT of untested code that's likely broken.

File: src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs (hunk 1 above)
Comment (filipnavara) [reply src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs:90]: Filed a bug: https://github.com/gssapi/gss-ntlmssp/issues/63

File: src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs (hunk 2 above)
Comment (filipnavara) [reply src/libraries/System.Net.Security/tests/UnitTests/NTAuthenticationTests.cs:89]: There's actually a bug in how the MIC is written into the authentication messages, so enabling it makes thing worse. I filed an issue: https://github.com/gssapi/gss-ntlmssp/issues/62

**Comment by wfurt** [comment]
I'm not sure if we really should go to the weeds of understanding NTLM internal in the tests. 
I would much rather expand "echo server" e.g. `corefx-net-http11.azurewebsites.net` to provide reference NTLM server and run functional tests agains it. 

cc: @rzikm @pavelsavara

**Comment by filipnavara** [comment]
> I'm not sure if we really should go to the weeds of understanding NTLM internal in the tests.

I tried to avoid it for as long as I could but I snapped in the end. There were too many issues not covered by the tests and that could not even easily be covered by the tests.

> I would much rather expand "echo server" e.g. corefx-net-http11.azurewebsites.net to provide reference NTLM server and run functional tests agains it.

If you are willing to host a test infrastructure I think that may provide a solid middle ground. However, the issue is not just HTTP. The SMTP GSSAPI authentication is currently broken on anything but Windows and there's no test coverage at all. `NegotiateStream` doesn't work on macOS and is not interoperable between systems (except maybe Linux and Windows if stars align, or if Kerberos is used).

We, at @emclient, have been hosting our own testing infrastructure for quite a while now and it's not easy. The Exchange servers bit rot constantly, the Windows Server 2019 machine on Azure needs to be restarted daily (seems to be a VM size issue and resource depletion).

I would be fine writing a simple echo SMTP server to cover the scenario above but I would need a commitment on hosting it and supporting it to be hammered by the .NET test infrastructure. Is that even something reasonable to ask?

**Comment by wfurt** [comment]
> I would be fine writing a simple echo SMTP server to cover the scenario above but I would need a commitment on hosting it and supporting it to be hammered by the .NET test infrastructure. Is that even something reasonable to ask?

probably. It would _finally_ close https://github.com/dotnet/runtime/issues/19436 after 5+ years of waiting ;(
Perhaps off-line chat with me and @karelz would be beneficial.

**Comment by filipnavara** [comment]
> Perhaps off-line chat with me and @karelz would be beneficial.

Definitely up for it. You know how to reach me.

**Comment by jstedfast** [comment]
Would love to know the conclusion of these discussions and whether [MailKit](https://github.com/jstedfast/MailKit) would be able to take advantage of this.

**Comment by wfurt** [comment]
Sorry for the delay @filipnava. Is this something you still want to push forward? It seems like RC4 was part of  #66879 so if we need it we should take it from there. 
I think we can take it even if various OS bugs will make it fragile IMHO.

**Comment by filipnavara** [comment]
> Is this something you still want to push forward?

Yes. I will rebase it and update the RC4 reference.

**Comment by filipnavara** [comment]
> Yes. I will rebase it and update the RC4 reference.

Done. The remaining test failure is unrelated (https://github.com/dotnet/runtime/issues/13757).

**Review by wfurt — APPROVED** [review]
LGTM.
Thanks @filipnavara

**Pull request merged**